Rock-Paper-Scissor Game

abdoalsir opened the pull request:
---
name: solution review for Rock-Paper-Scissor game
about: A template PR for code review with a checklist for the Rock-Paper-Scissor game
---

<!--
  make this PR easy to find:

  - assign: yourself
  - link it to the corresponding challenge issue on the project board
-->

<!-- describe the challenge you will solve -->

## Behavior

### Files

- [X] The file name describes the function's behavior
- [X] There is a module docstring in the function file
- [X] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [X] There is a module docstring in the tests file

### Unit Tests

- [X] The test class has a helpful name in PascalCase
- [X] The test class has a docstring
- Every unit test has
  - [X] A helpful name
  - [X] A clear docstring
  - [X] Only one assertion
  - [X] There is no logic in the unit test
- [X] All tests pass
- [X] There are tests for defensive assertions
- [X] There are tests for boundary cases

### Function Docstring

- [X] The function's behavior is described
- The function's arguments are described:
  - [X] Type
  - [X] Purpose
  - [X] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [X] Type
  - [X] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [X] Each assumption about an argument is checked with an assertion
  - [X] Each assertion checks for _only one_ assumption about the argument
- [X] Include 3 or more (passing!) doctests

### The Function

- [X] The function's name describes it's behavior
- [X] The function's name matches the file name
- [X] The function has correct type annotations
- [X] The function is not called in the function file

## Strategy

### Do's

- [X] Variable names help to understand the strategy
- [X] Any comments are clear and describe the strategy
- [X] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [X] The function's strategy _is not_ described in the documentation
- [X] Comments explain the _strategy_, **not** the _implementation_
- [X] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [X] The code passes the formatting checks
- [X] The code passes all Ruff linting checks
- [X] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [X] Variables are named with snake_case
- [X] Variable names are clear and helpful
- [X] The code follows the strategy as simply as possible
- [X] The implementation is as simple as possible given the strategy
- [X] There are no commented lines of code
- [X] There are no `print` statements anywhere
- [X] The code includes defensive assertions
- [X] Defensive assertions include as little logic as possible

Aseel-AbuKmail reviewed (commented):
Hi Abdulrahman!
I thoroughly enjoyed reviewing your challenge and would like to offer a few suggestions as feedback:

For Rock-Paper-Scissors Module:

play_round function (line 17):
The function is well-documented and easy to follow. You might consider enhancing the user experience by returning both the result and the computer's choice. For instance, instead of just 'win', the function could return ('win', 'scissors') to show what the computer picked.

options list (line 14):
Currently, options is a mutable list. To indicate that it is not meant to be changed, you could define it as a constant (OPTIONS = [...]) and use uppercase naming conventions as per Python standards.

General Suggestion:
You could include type hints for the options variable (e.g., options: list[str] = [...]) to make the code more type-safe and readable.

Aseel-AbuKmail reviewed (commented):
For Test Suite

test_rock method (line 17):
Overriding random.choice is an effective way to ensure deterministic outcomes. However, you should restore the original behavior of random.choice after the test to avoid unintended side effects. Using the unittest.mock library for patching would make this cleaner and safer.

test_case_insensitivity method (line 37):
Good job testing for case insensitivity! This ensures the function is more user-friendly.

test_empty_input method (line 47):
The test for empty input is great. To further enhance coverage, consider adding tests for invalid types (e.g., integers, floats, or lists).

Import Path:
The relative import (from ..rock_paper_scissor) might cause issues if the test suite is run independently. An absolute import could make the module easier to test in isolation.

abdoalsir commented:
Thank you Aseel for review, I will make these changes and push the updated code

abdoalsir commented:
Hi @Aseel-AbuKmail,
I’ve implemented the suggestions you provided and updated the code accordingly. Could you please complete the review? If everything looks good and you have no further suggestions, feel free to merge the PR into the main branch.
Let me know if there’s anything else that needs improvement!

Aseel-AbuKmail commented:
> Hi @Aseel-AbuKmail, I’ve implemented the suggestions you provided and updated the code accordingly. Could you please complete the review? If everything looks good and you have no further suggestions, feel free to merge the PR into the main branch. Let me know if there’s anything else that needs improvement!

Thank you for taking my feedback into account. I’ve reviewed the updates, and I will merge your pull request into the main branch and delete the challenge branch.
